src/replication.c

```diff
@@ -1975,6 +1975,22 @@ void readSyncBulkPayload(connection *conn) {
     /* We need to stop any AOF rewriting child before flushing and parsing
      * the RDB, otherwise we'll create a copy-on-write disaster. */
     if (server.aof_state != AOF_OFF) stopAppendOnly();
+    /* Also try to stop save RDB child before flushing and parsing thr RDB:
+     * 1. avoid RDB file race if load the RDB on disk.
+     * 2. avoid create a copy-on-write disaster, too. */
+    if (server.child_type == CHILD_TYPE_RDB &&
+        (!use_diskless_load || server.repl_diskless_load == REPL_DISKLESS_LOAD_SWAPDB))
```

Comment:
I'm not sure we want that second line in the condition...
as you pointed out, we have 3 scenarios:
1. disk-based replication, in which we want to avoid accidentally overwriting the mew file. and also as you pointed we rather do that before emptyData in order to avoid massive COW.
2. diskless without swapdb, here too we call emptyData so we want to avoid massive COW (no concern about overwriting the file here)
3. diskless swapdb, in which we don't do emptyData, so no massive COW, but we do consume a lot of memory so we may want to release even a small COW we can.

you're avoiding the kill on the 2nd scenario, right? (not sure why), am i missing anything?

Reply:
On the 2nd scenario, when receiving RDB from the master fails (maybe master core dump), but we call emptyData, resulting in all data loss, so we avoid the kill to ensure data integrity, even if it may cause COW. WDYT?

Reply:
if that were a legitimate solution to that concern, we would have had to make sure there's a rewrite triggered just before we flush.
with the current state, all the code here is just for the rare chance there is some bgsave fork already running. we can't argue that we're relying on it.

bottom line, i think we just need to move the code as you did, but not add any conditions to it.

Reply:
I agree with you and delete second line in the condition. Thx.